File: FineraApp/backend/FineraAPI/Models/Transaction.cs
//Models/Transaction.cs

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace FineraAPI.Models
{
    public class Transaction
    {
        public int Id { get; set; }
        
        [Required]
        public int UserId { get; set; }
        
        [Required]
        public int CategoryId { get; set; }
        
        [Required]
        [Column(TypeName = "decimal(10,2)")]
        public decimal Amount { get; set; }
        
        [StringLength(255)]
        public string? Description { get; set; }
        
        [Required]
        public DateTime TransactionDate { get; set; }
        
        [Required]
        public string Type { get; set; } = string.Empty; // Income or Expense
        
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
        
        // Navigation properties
        public virtual User User { get; set; } = null!;
        public virtual Category Category { get; set; } = null!;
    }
}

File: FineraApp/frontend/finera_flutter_app/lib/models/transaction.dart
//models/transaction.dart

class TransactionModel {
  final int id;
  final int categoryId;
  final String categoryName;
  final String categoryType; // "Income" or "Expense" (from category)
  final double amount;
  final String? description;
  final DateTime transactionDate;
  final String type; // "Income" or "Expense" (explicit type)

  TransactionModel({
    required this.id,
    required this.categoryId,
    required this.categoryName,
    required this.categoryType,
    required this.amount,
    this.description,
    required this.transactionDate,
    required this.type,
  });

  factory TransactionModel.fromJson(Map<String, dynamic> json) {
    return TransactionModel(
      id: json['id'] as int,
      categoryId: json['categoryId'] as int,
      categoryName: (json['categoryName'] ?? '') as String,
      categoryType: (json['categoryType'] ?? '') as String,
      amount: (json['amount'] as num).toDouble(),
      description: json['description'] as String?,
      transactionDate: DateTime.parse(json['transactionDate'] as String),
      type: (json['type'] ?? 'Expense') as String,
    );
  }

  Map<String, dynamic> toJsonCreate() {
    return {
      'categoryId': categoryId,
      'amount': amount,
      'description': description,
      'transactionDate': transactionDate.toIso8601String(),
      'type': type,
    };
  }

  Map<String, dynamic> toJsonUpdate() {
    return {
      'categoryId': categoryId,
      'amount': amount,
      'description': description,
      'transactionDate': transactionDate.toIso8601String(),
      'type': type,
    };
  }
}

